pkg/services/ngalert/notifier/channels/slack.go

```diff
@@ -0,0 +1,247 @@
+		tmpl:           t,
+		externalUrl:    externalUrl,
+	}, nil
+}
+
+// slackMessage is the slackMessage for sending a slack notification.
+type slackMessage struct {
+	Channel     string                   `json:"channel,omitempty"`
+	Username    string                   `json:"username,omitempty"`
+	IconEmoji   string                   `json:"icon_emoji,omitempty"`
+	IconURL     string                   `json:"icon_url,omitempty"`
+	Attachments []attachment             `json:"attachments"`
+	Blocks      []map[string]interface{} `json:"blocks"`
+}
+
+// attachment is used to display a richly-formatted message block.
+type attachment struct {
+	Title      string              `json:"title,omitempty"`
+	TitleLink  string              `json:"title_link,omitempty"`
+	Text       string              `json:"text"`
+	Fallback   string              `json:"fallback"`
+	Fields     []config.SlackField `json:"fields,omitempty"`
+	Footer     string              `json:"footer"`
+	FooterIcon string              `json:"footer_icon"`
+	Color      string              `json:"color,omitempty"`
+	Ts         int64               `json:"ts,omitempty"`
+}
+
+const (
+	ColorAlertFiring = "#D63232"
+	ColorAlertResolved
+)
+
+func getAlertStatusColor(status model.AlertStatus) string {
+	if status == model.AlertFiring {
+		return ColorAlertFiring
+	}
+	return ColorAlertResolved
+}
+
+func (sn *SlackNotifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
+	msg, err := sn.buildSlackMessage(ctx, as)
+	if err != nil {
+		return false, errors.Wrap(err, "build slack message")
+	}
+
+	b, err := json.Marshal(msg)
+	if err != nil {
+		return false, errors.Wrap(err, "marshal json")
+	}
+
+	cmd := &models.SendWebhookSync{
+		Url:        sn.URL,
+		Body:       string(b),
+		HttpMethod: http.MethodPost,
+	}
+
+	if sn.Token != "" {
+		sn.log.Debug("Adding authorization header to HTTP request")
+		cmd.HttpHeader = map[string]string{
+			"Authorization": fmt.Sprintf("Bearer %s", sn.Token),
+		}
+	}
+
+	if err := bus.DispatchCtx(ctx, cmd); err != nil {
+		sn.log.Error("Failed to send slack notification", "error", err, "webhook", sn.Name)
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (sn *SlackNotifier) buildSlackMessage(ctx context.Context, as []*types.Alert) (*slackMessage, error) {
+	var tmplErr error
+	data := notify.GetTemplateData(ctx, &template.Template{ExternalURL: sn.externalUrl}, as, gokit_log.NewNopLogger())
+	alerts := types.Alerts(as...)
+	tmpl := notify.TmplText(sn.tmpl, data, &tmplErr)
+
+	req := &slackMessage{
+		Channel:   tmpl(sn.Recipient),
+		Username:  tmpl(sn.Username),
+		IconEmoji: tmpl(sn.IconEmoji),
+		IconURL:   tmpl(sn.IconURL),
+		Attachments: []attachment{
+			{
+				Color:      getAlertStatusColor(alerts.Status()),
+				Title:      tmpl(sn.Title),
+				Fallback:   tmpl(sn.Fallback),
+				Footer:     "Grafana v" + setting.BuildVersion,
+				FooterIcon: "https://grafana.com/assets/img/fav32.png",
```

Comment:
should this also be a constant?

Reply:
I think this should be fine here. If I find it being used in multiple places, I will make it a constant.

Reply:
Or let me just make it a constant :)